autonomi/src/client/data_map_restoration.rs

```diff
@@ -0,0 +1,150 @@
+use crate::client::{Client, GetError};
+use crate::self_encryption::DataMapLevel;
+use bytes::Bytes;
+use eyre::Result;
+use self_encryption::{ChunkInfo, DataMap, get_root_data_map};
+use xor_name::XorName;
+
+impl Client {
+    /// Restore a complete data map from a DataMapChunk, handling both old and new formats
+    /// This function properly handles recursive DataMapLevel schemes by traversing all levels
+    /// until it reaches the root data map containing the actual file chunks.
+    ///
+    /// self_encryption now changed to always return a data_map pointing to the 3 datamap_chunks.
+    /// Hence, the input data_map_bytes is actually the root_data_map pointing to that 3 datamap_chunks.
+    /// The downloading work flow then shall be:
+    ///   * fetch that 3 datamap_chunks first
+    ///   * compose the real datamap of the file, from that 3 datamap_chunks
+    ///   * repeat the above two steps if the recovered data_map contains chiled
+    ///   * fetch the leftover content chunks to compose the file
+    pub(crate) async fn restore_data_map_from_chunk(
+        &self,
+        data_map_chunk: &DataMapChunk,
+    ) -> Result<DataMap, GetError> {
+        let mut data_map_bytes = data_map_chunk.0.value().clone();
+
+        // In case the input is the new data_map root, restore it further,
+        // before going into the further recursive.
+        if let Ok(data_map) = rmp_serde::from_slice::<DataMap>(&data_map_bytes) {
+            info!("Restoring from new root data_map:\n{data_map:?}");
+            let file_data_map = self.fetch_new_data_map(&data_map)?;
+
+            info!("Fetched file data_map of new version: \n{file_data_map:?}");
+            return Ok(file_data_map);
+        }
+
+        loop {
+            // The data_map_bytes could be an Archive, we shall return earlier for that case
+            match Self::deserialize_data_map(&data_map_bytes) {
+                Ok(mut data_map) => {
+                    info!("Restoring from data_map:\n{data_map:?}");
+                    if !data_map.is_child() {
+                        return Ok(data_map);
+                    }
+                    data_map.child = None;
+                    data_map_bytes = self.fetch_from_data_map(&data_map).await?;
+                }
+                Err(e) => {
+                    info!("Failed to deserialize data_map_bytes: {e:?}");
+                    return Err(GetError::Configuration(format!("{e:?}")));
+                }
+            }
+        }
+    }
+
+    /// Fetch the file data_map from the root one using lazy evaluation.
+    /// Chunks are only fetched from the network when actually needed by get_root_data_map.
+    fn fetch_new_data_map(&self, data_map: &DataMap) -> Result<DataMap, GetError> {
+        let total_chunks = data_map.infos().len();
+        #[cfg(feature = "loud")]
+        println!(
+            "Setting up lazy chunk fetching for {total_chunks} chunks of data_map {data_map:?}"
```

Comment:
I would suggest: "Using lazy chunk fetching for {total_chunks} of datamap {data_map:?}"